File: IdentityX/testing_framework/verification_test.go
package testing

import (
	"GoAuth/internal/apierr"
	"encoding/json"
	"errors"
	"net/http"
	"net/url"
	"regexp"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
)

func testVerification(t *testing.T, suite *TestSuite) {
	client := suite.NewClient(t)
	user := client.WithCredentials("verification@mail.com", ValidPassword).Register().Login().CreateProject("verification")

	time.Sleep(time.Millisecond * 100)

	link, err := GetLatestVerificationLink()
	if err != nil {
		t.Fatalf("Failed getting verification link: %v", err.Error())
	}

	if link == "" {
		t.Fatalf("Verification link is empty")
	}

	u, err := url.Parse(link)
	if err != nil {
		t.Fatalf("Failed parsing verification link: %v", err)
	}

	t.Run("ResendVerificationEmail", func(t *testing.T) {
		user.WithT(t).POST("/auth/verify/resend").
			Expect(http.StatusOK).
			HasMessage("verification email resent successfully")

		time.Sleep(time.Millisecond * 100)

		link2, err2 := GetLatestVerificationLink()
		if err2 != nil {
			t.Fatalf("Failed getting verification link: %v", err2.Error())
		}

		if link2 == "" {
			t.Fatalf("Verification link is empty")
		}

		u2, err2 := url.Parse(link2)
		if err2 != nil {
			t.Fatalf("Failed parsing verification link: %v", err2)
		}

		assert.NotEqualf(t, u2.Query().Get("token"), u.Query().Get("token"), "tokens shouldn't be equal")
	})

	t.Run("VerifyUser", func(t *testing.T) {
		user.WithT(t).POST(u.Path).
			WithQuery("token", u.Query().Get("token")).
			Expect(http.StatusOK).
			HasMessage("user verified, please refresh")
	})

	t.Run("VerifyUserAgain", func(t *testing.T) {
		user.WithT(t).POST(u.Path).
			WithQuery("token", u.Query().Get("token")).
			Expect(http.StatusOK).
			HasMessage("user verified, please refresh")
	})

	t.Run("ResendVerificationEmailNotAllowed", func(t *testing.T) {
		user.WithT(t).POST("/auth/verify/resend").
			Expect(http.StatusForbidden).
			HasErrID(apierr.AuthAlreadyVerified).
			HasMessage("user already verified")
	})

	t.Run("SessionInfoBeforeRefreshed", func(t *testing.T) {
		user := suite.NewClient(t).WithAuth(user.auth)

		data := user.GET("/sessions/me").
			Expect(http.StatusOK).
			RequireDataValue()

		spec := map[string]interface{}{
			"refresh_expire_date": AnyNumber{},
			"access": map[string]interface{}{
				"iss": "GoAuth",
				"exp": AnyNumber{},
				"iat": AnyNumber{},
				"jti": AnyUUID{},
				"sub": map[string]interface{}{
					"id":          AnyUUID{},
					"email":       "verification@mail.com",
					"project_id":  nil,
					"user_type":   "client",
					"metadata":    nil,
					"session_id":  AnyUUID{},
					"user_agent":  AnyString{},
					"user_ip":     AnyString{},
					"is_verified": false,
					"verified_at": nil,
				},
			},
		}

		Validate(t, data, spec)
	})

	t.Run("SessionInfoRefreshed", func(t *testing.T) {
		user := suite.NewClient(t).WithCredentials("verification@mail.com", ValidPassword).Login()

		data := user.GET("/sessions/me").
			Expect(http.StatusOK).
			RequireDataValue()

		spec := map[string]interface{}{
			"refresh_expire_date": AnyNumber{},
			"access": map[string]interface{}{
				"iss": "GoAuth",
				"exp": AnyNumber{},
				"iat": AnyNumber{},
				"jti": AnyUUID{},
				"sub": map[string]interface{}{
					"id":          AnyUUID{},
					"email":       "verification@mail.com",
					"project_id":  nil,
					"user_type":   "client",
					"metadata":    nil,
					"session_id":  AnyUUID{},
					"user_agent":  AnyString{},
					"user_ip":     AnyString{},
					"is_verified": true,
					"verified_at": AnyDate{},
				},
			},
		}

		Validate(t, data, spec)
	})

	var projectUser *Client
	projectUser = client.WithCredentials("verification-project@mail.com", ValidPassword).ProjectRegister(user.projectID).ProjectLogin(user.projectID)

	time.Sleep(time.Millisecond * 100)

	link2, err := GetLatestVerificationLink()
	if err != nil {
		t.Fatalf("Failed getting verification link: %v", err.Error())
	}

	if link2 == "" {
		t.Fatalf("Verification link is empty")
	}

	u2, err := url.Parse(link2)
	if err != nil {
		t.Fatalf("Failed parsing verification link: %v", err)
	}

	t.Run("ResendVerificationEmailProjectUser", func(t *testing.T) {
		projectUser.WithT(t).POST("/auth/verify/resend").
			Expect(http.StatusOK).
			HasMessage("verification email resent successfully")

		time.Sleep(time.Millisecond * 100)

		link3, err2 := GetLatestVerificationLink()
		if err2 != nil {
			t.Fatalf("Failed getting verification link: %v", err2.Error())
		}

		if link3 == "" {
			t.Fatalf("Verification link is empty")
		}

		u3, err2 := url.Parse(link3)
		if err2 != nil {
			t.Fatalf("Failed parsing verification link: %v", err2)
		}

		assert.NotEqualf(t, u3.Query().Get("token"), u2.Query().Get("token"), "tokens shouldn't be equal")
	})

	t.Run("VerifyProjectUser", func(t *testing.T) {
		projectUser.WithT(t).POST(u2.Path).
			WithQuery("token", u2.Query().Get("token")).
			Expect(http.StatusOK).
			HasMessage("user verified, please refresh")
	})

	t.Run("VerifyProjectUserAgain", func(t *testing.T) {
		projectUser.WithT(t).POST(u2.Path).
			WithQuery("token", u2.Query().Get("token")).
			Expect(http.StatusOK).
			HasMessage("user verified, please refresh")
	})

	t.Run("ResendVerificationEmailNotAllowedProjectUser", func(t *testing.T) {
		projectUser.WithT(t).POST("/auth/verify/resend").
			Expect(http.StatusForbidden).
			HasErrID(apierr.AuthAlreadyVerified).
			HasMessage("user already verified")
	})

	t.Run("ProjUserSessionInfoBeforeRefreshed", func(t *testing.T) {
		data := projectUser.WithT(t).GET("/sessions/me").
			Expect(http.StatusOK).
			RequireDataValue()

		spec := map[string]interface{}{
			"refresh_expire_date": AnyNumber{},
			"access": map[string]interface{}{
				"iss": "GoAuth",
				"exp": AnyNumber{},
				"iat": AnyNumber{},
				"jti": AnyUUID{},
				"sub": map[string]interface{}{
					"id":          AnyUUID{},
					"email":       "verification-project@mail.com",
					"project_id":  AsString{Value: user.projectID, Matcher: AnyUUID{}},
					"user_type":   "project",
					"metadata":    map[string]interface{}{},
					"session_id":  AnyUUID{},
					"user_agent":  AnyString{},
					"user_ip":     AnyString{},
					"is_verified": false,
					"verified_at": nil,
				},
			},
		}

		Validate(t, data, spec)
	})

	t.Run("ProjUserSessionInfoRefreshed", func(t *testing.T) {
		projectUser := suite.NewClient(t).WithCredentials("verification-project@mail.com", ValidPassword).ProjectLogin(user.projectID)

		data := projectUser.GET("/sessions/me").
			Expect(http.StatusOK).
			RequireDataValue()

		spec := map[string]interface{}{
			"refresh_expire_date": AnyNumber{},
			"access": map[string]interface{}{
				"iss": "GoAuth",
				"exp": AnyNumber{},
				"iat": AnyNumber{},
				"jti": AnyUUID{},
				"sub": map[string]interface{}{
					"id":          AnyUUID{},
					"email":       "verification-project@mail.com",
					"project_id":  AsString{Value: user.projectID, Matcher: AnyUUID{}},
					"user_type":   "project",
					"metadata":    map[string]interface{}{},
					"session_id":  AnyUUID{},
					"user_agent":  AnyString{},
					"user_ip":     AnyString{},
					"is_verified": true,
					"verified_at": AnyDate{},
				},
			},
		}

		Validate(t, data, spec)
	})
}

type MailHogResponse struct {
	Items []struct {
		Content struct {
			Body string `json:"Body"`
		} `json:"Content"`
	} `json:"items"`
}

func GetLatestVerificationLink() (string, error) {
	resp, err := http.Get("http://mailhog:8025/api/v2/messages")
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var mh MailHogResponse
	if err := json.NewDecoder(resp.Body).Decode(&mh); err != nil {
		return "", err
	}

	if len(mh.Items) == 0 {
		return "", errors.New("no emails found")
	}

	re := regexp.MustCompile(`href="([^"]+)"`)
	match := re.FindStringSubmatch(mh.Items[0].Content.Body)
	if len(match) < 2 {
		return "", errors.New("verification link not found")
	}

	return match[1], nil
}

File: IdentityX/testing_framework/verification.test.ts
import { beforeAll, describe, expect, test } from "vitest";
import { createClient } from "./helpers/index.js";
import { assertErrID, assertMessage, shouldFail } from "./helpers/assert.js";
import {
    Validate,
    AnyUUID,
    AnyDate,
    AnyNumber,
    AnyString,
    AsString,
} from "./helpers/validate.js";
import { ValidPassword } from "./fixtures/auth/testdata.js";

// AuthAlreadyVerified = fail.ID(0, "AUTH", 4, true, ...) → 0_AUTH_0004_S
const ErrAuthAlreadyVerified = "0_AUTH_0004_S";

const MAILHOG_URL = process.env.MAILHOG_URL ?? "http://localhost:8025";

interface MailHogResponse {
    items: Array<{
        Content: {
            Body: string;
        };
    }>;
}

async function getLatestVerificationLink(): Promise<string> {
    const res = await fetch(`${MAILHOG_URL}/api/v2/messages`);
    if (!res.ok) throw new Error(`MailHog request failed: ${res.status}`);

    const mh: MailHogResponse = await res.json();
    if (!mh.items || mh.items.length === 0) throw new Error("no emails found");

    const body = mh.items[0].Content.Body;
    const match = body.match(/href="([^"]+)"/);
    if (!match || match.length < 2) throw new Error("verification link not found in email");

    return match[1];
}

function sleep(ms: number) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}

describe("Verification", () => {
    let user: any;
    let projectID: string;
    let initialLink: URL;

    beforeAll(async () => {
        await createClient().withCredentials("verification@mail.com", ValidPassword).register();
        user = await createClient().withCredentials("verification@mail.com", ValidPassword).login();
        const project = await user.post("/projects", { project_name: "verification", metadata: { env: "test" } });
        projectID = project.id;

        await sleep(100);
        const link = await getLatestVerificationLink();
        if (!link) throw new Error("Verification link is empty");
        initialLink = new URL(link);
    });

    test("ResendVerificationEmail", async () => {
        await user.post("/auth/verify/resend");

        await sleep(100);
        const link2 = await getLatestVerificationLink();
        if (!link2) throw new Error("Verification link is empty after resend");

        const u2 = new URL(link2);
        expect(u2.searchParams.get("token")).not.toBe(initialLink.searchParams.get("token"));
    });

    test("VerifyUser", async () => {
        const path = initialLink.pathname;
        const token = initialLink.searchParams.get("token");
        const res = await user.http.post(`${path}?token=${token}`);
        expect(res.data?.message ?? res.data?.data?.message).toMatch(/user verified/i);
    });

    test("VerifyUserAgain", async () => {
        // Same token — server should still return 200
        const path = initialLink.pathname;
        const token = initialLink.searchParams.get("token");
        const res = await user.http.post(`${path}?token=${token}`);
        expect(res.data?.message ?? res.data?.data?.message).toMatch(/user verified/i);
    });

    test("ResendVerificationEmailNotAllowed", async () => {
        const body = await shouldFail(user.post("/auth/verify/resend"), 403);
        assertErrID(body, ErrAuthAlreadyVerified);
        assertMessage(body, "user already verified");
    });

    test("SessionInfoBeforeRefreshed", async () => {
        // Use the original auth cookies (token still says is_verified: false)
        const data = await user.get("/sessions/me");
        Validate(data, {
            refresh_expire_date: AnyNumber,
            access: {
                iss: "GoAuth",
                exp: AnyNumber,
                iat: AnyNumber,
                jti: AnyUUID,
                sub: {
                    id:          AnyUUID,
                    email:       "verification@mail.com",
                    project_id:  null,
                    user_type:   "client",
                    metadata:    null,
                    session_id:  AnyUUID,
                    user_agent:  AnyString,
                    user_ip:     AnyString,
                    is_verified: false,
                    verified_at: null,
                },
            },
        });
    });

    test("SessionInfoRefreshed", async () => {
        // Fresh login — token now reflects verified state
        const freshUser = await createClient()
            .withCredentials("verification@mail.com", ValidPassword)
            .login();

        const data = await freshUser.get("/sessions/me");
        Validate(data, {
            refresh_expire_date: AnyNumber,
            access: {
                iss: "GoAuth",
                exp: AnyNumber,
                iat: AnyNumber,
                jti: AnyUUID,
                sub: {
                    id:          AnyUUID,
                    email:       "verification@mail.com",
                    project_id:  null,
                    user_type:   "client",
                    metadata:    null,
                    session_id:  AnyUUID,
                    user_agent:  AnyString,
                    user_ip:     AnyString,
                    is_verified: true,
                    verified_at: AnyDate,
                },
            },
        });
    });

    // ── Project user verification ──────────────────────────────────────────────

    describe("ProjectUser", () => {
        let projectUser: any;
        let projInitialLink: URL;

        beforeAll(async () => {
            // Register + login as project user
            await createClient().http.post(`/projects/${projectID}/register`, {
                email: "verification-project@mail.com",
                password: ValidPassword,
            });
            projectUser = await createClient()
                .withCredentials("verification-project@mail.com", ValidPassword)
                .projectLogin(projectID);

            await sleep(100);
            const link = await getLatestVerificationLink();
            if (!link) throw new Error("Project user verification link is empty");
            projInitialLink = new URL(link);
        });

        test("ResendVerificationEmailProjectUser", async () => {
            await projectUser.post("/auth/verify/resend");

            await sleep(100);
            const link3 = await getLatestVerificationLink();
            if (!link3) throw new Error("Project user verification link is empty after resend");

            const u3 = new URL(link3);
            expect(u3.searchParams.get("token")).not.toBe(projInitialLink.searchParams.get("token"));
        });

        test("VerifyProjectUser", async () => {
            const path = projInitialLink.pathname;
            const token = projInitialLink.searchParams.get("token");
            const res = await projectUser.http.post(`${path}?token=${token}`);
            expect(res.data?.message ?? res.data?.data?.message).toMatch(/user verified/i);
        });

        test("VerifyProjectUserAgain", async () => {
            const path = projInitialLink.pathname;
            const token = projInitialLink.searchParams.get("token");
            const res = await projectUser.http.post(`${path}?token=${token}`);
            expect(res.data?.message ?? res.data?.data?.message).toMatch(/user verified/i);
        });

        test("ResendVerificationEmailNotAllowedProjectUser", async () => {
            const body = await shouldFail(projectUser.post("/auth/verify/resend"), 403);
            assertErrID(body, ErrAuthAlreadyVerified);
            assertMessage(body, "user already verified");
        });

        test("ProjUserSessionInfoBeforeRefreshed", async () => {
            const data = await projectUser.get("/sessions/me");
            Validate(data, {
                refresh_expire_date: AnyNumber,
                access: {
                    iss: AsString(projectID, AnyUUID),
                    exp: AnyNumber,
                    iat: AnyNumber,
                    jti: AnyUUID,
                    sub: {
                        id:          AnyUUID,
                        email:       "verification-project@mail.com",
                        project_id:  AsString(projectID, AnyUUID),
                        user_type:   "project",
                        metadata:    {},
                        session_id:  AnyUUID,
                        user_agent:  AnyString,
                        user_ip:     AnyString,
                        is_verified: false,
                        verified_at: null,
                    },
                },
            });
        });

        test("ProjUserSessionInfoRefreshed", async () => {
            const freshProjUser = await createClient()
                .withCredentials("verification-project@mail.com", ValidPassword)
                .projectLogin(projectID);

            const data = await freshProjUser.get("/sessions/me");
            Validate(data, {
                refresh_expire_date: AnyNumber,
                access: {
                    iss: AsString(projectID, AnyUUID),
                    exp: AnyNumber,
                    iat: AnyNumber,
                    jti: AnyUUID,
                    sub: {
                        id:          AnyUUID,
                        email:       "verification-project@mail.com",
                        project_id:  AsString(projectID, AnyUUID),
                        user_type:   "project",
                        metadata:    {},
                        session_id:  AnyUUID,
                        user_agent:  AnyString,
                        user_ip:     AnyString,
                        is_verified: true,
                        verified_at: AnyDate,
                    },
                },
            });
        });
    });
});